### PR TITLE
ChannelSummation feature branch

### PR DESCRIPTION
**Pull request by psurukuc**
Pull request to request for a quick review and comment on the naming/coding standards. 
Please let me know if the standards are being followed and I will change according to the comments. 
Warning: Not ready for merging into the 'develop' branch yet

### CODE REVIEW

**Comment by pslocum** [comment]
Thanks for committing this.  To summarize, would you say that the present status is that the channel PSDs are summed, and that the next step is to write the new summed object to a root file?  Also, a comment on this comment:     
**- "ps-ch": void (Nymph::KTDataPtr) -- Adds channels PSD now, will be moved to addition in frequencies; Requires KTPowerSpectrumData; Adds summation of the channel results; Emits signal "ps"**
We are summing the PSD across frequency now, and will keep doing that going forward.  The next step will be to use fftw-phase while we are doing that.

**Comment by psurukuc** [comment]
Penny,

That is right. Currently, channel PSDs are summed and the next step is to
write the new object to a root file.

About the comment, sorry, you are right, I will fix the comment.

Pranava

File: Source/SpectrumAnalysis/KTChannelAggregator.hh [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    // Get the grid size assuming a square grid
+    // Set default to 30 currently
+    int fNGrid = 30;
+    
+    // PTS:: for 18.6 keV electrons, this somehow needs to come from the data file or config file
+    double fWavelength=0.0115;
+    
+    //For exception handling to make sure the grid is defined before the spectra are assigned.
+    bool fIsGridDefined=false;
+
```
Comment (nsoblath) [inline Source/SpectrumAnalysis/KTChannelAggregator.hh]: Name variables in function definitions to make argument meanings clear from the header alone

File: Examples/ConfigFiles/KatydidPSConfig.yaml [hunk 2]
```diff
@@ -16,8 +16,8 @@ processor-toolbox:
         slot: "fft:ts-fftw"
       - signal: "fft:fft"
         slot: "to-ps:fs-fftw-to-psd"
-      - signal: "to-ps:psd"
-        slot: "writer:ps"
+      - signal: "to-ps:ps"
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Examples/ConfigFiles/KatydidPSConfig.yaml]: You are right, removed the change in the file.

[the remark this answers is not in the document]
Comment (psurukuc) [reply Examples/ConfigFiles/KatydidPSConfig.yaml]: Fixed in commit 77127bbbb44983cddd569541ee0bf3ba2dad9102

File: Source/Data/SpectrumAnalysis/KTChannelAggregatedData.cc [hunk 3]
```diff
@@ -0,0 +1,101 @@
+/*
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/Data/SpectrumAnalysis/KTChannelAggregatedData.cc]: Fixed in commit 77127bb

File: Source/Data/SpectrumAnalysis/KTChannelAggregatedData.hh [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+      double fGridPointX;
+      
+      /// The Y position of the grid point
+      double fGridPointY;
+      
+      /// Magnitude at the defined grid location
+      double fMagnitude;
+    };
+    typedef std::vector<KTGrid> SetOfGridPoints;
+
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/Data/SpectrumAnalysis/KTChannelAggregatedData.hh]: Fixed in commit 77127bb

File: Source/IO/BasicROOTFileWriter/KTBasicROOTTypeWriterSpectrumAnalysis.cc [hunk 5]
```diff
@@ -23,6 +23,7 @@
 #include "KTTimeFrequencyPolar.hh"
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/IO/BasicROOTFileWriter/KTBasicROOTTypeWriterSpectrumAnalysis.cc]: Fixed in commit 77127bb

File: Source/IO/Conversions/KT2ROOT.cc [hunk 6]
```diff
@@ -25,6 +25,8 @@
 #include "KTTimeSeriesReal.hh"
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/IO/Conversions/KT2ROOT.cc]: Fixed in commit 77127bb

File: Source/IO/ROOTSpectrogramWriter/KTROOTSpectrogramWriter.hh [hunk 7]
```diff
@@ -252,6 +259,88 @@ namespace Katydid
 
          return;
      }
+
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/IO/ROOTSpectrogramWriter/KTROOTSpectrogramWriter.hh]: Fixed in commit 77127bb

File: Source/SpectrumAnalysis/KTAggregatedChannelOptimizer.cc [hunk 8]
```diff
@@ -0,0 +1,64 @@
+/*
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/SpectrumAnalysis/KTAggregatedChannelOptimizer.cc]: Fixed in commit 77127bb

File: Source/Data/SpectrumAnalysis/KTChannelAggregatedData.hh [hunk 9]
```diff
@@ -0,0 +1,230 @@
+/**
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/Data/SpectrumAnalysis/KTChannelAggregatedData.hh]: Fixed in commit 77127bb

File: Source/IO/BasicROOTFileWriter/KTBasicROOTTypeWriterSpectrumAnalysis.hh [hunk 10]
```diff
@@ -88,6 +88,18 @@ namespace Katydid
             //************************
         public:
             void WriteKDTreeSparseSpectrogram(Nymph::KTDataPtr data);
+      
+      
+      //************************
+      // Aggregated Channel Data
+      //************************
+    public:
+            void WriteAggregatedFrequencySpectrumFFTWData(Nymph::KTDataPtr data);
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/IO/BasicROOTFileWriter/KTBasicROOTTypeWriterSpectrumAnalysis.hh]: Fixed in commit 77127bb

File: Source/IO/ROOTSpectrogramWriter/KTROOTSpectrogramTypeWriterSpectrumAnalysis.cc [hunk 11]
```diff
@@ -0,0 +1,100 @@
+/*
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/IO/ROOTSpectrogramWriter/KTROOTSpectrogramTypeWriterSpectrumAnalysis.cc]: Fixed in commit 77127bb

File: Source/SpectrumAnalysis/KTAggregatedChannelOptimizer.cc [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+#include "KTAggregatedChannelOptimizer.hh"
+
+#include "KTLogger.hh"
+
+namespace Katydid
+{
+  KTLOGGER(agglog, "KTAggregatedChannelOptimizer");
+  
+  // Register the processor
+  KT_REGISTER_PROCESSOR(KTAggregatedChannelOptimizer, "channelsum-optimizer");
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/SpectrumAnalysis/KTAggregatedChannelOptimizer.cc]: Fixed in commit 77127bb

File: Source/SpectrumAnalysis/KTAggregatedChannelOptimizer.cc [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+  {
+    const KTFrequencySpectrumFFTW* freqSpectrum=aggData.GetSpectrumFFTW(0);
+    // Get the number of frequency bins from the first component of aggData
+    int nFreqBins=freqSpectrum->GetNFrequencyBins();
+    int nGridPoints=aggData.GetNComponents(); // Get number of components
+    
+    int maxGridPoint=0.0;
+    double maxVoltage=0.0;
+    // Loop over each component and find and save the maximum summed absolute voltages and the corresponding frequencies
+    for (unsigned iGridPoint=0; iGridPoint<nGridPoints; ++iGridPoint){
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/SpectrumAnalysis/KTAggregatedChannelOptimizer.cc]: Fixed in commit 77127bb

File: Source/SpectrumAnalysis/KTAggregatedChannelOptimizer.cc [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+      
+      double gridLocationX;
+      double gridLocationY;
+      freqSpectrum=aggData.GetSpectrumFFTW(iGridPoint);
+      aggData.GetGridPoint(iGridPoint,gridLocationX,gridLocationY);
+      double maxVoltageFreq=aggData.GetSummedGridVoltage(iGridPoint);
+
+      if(maxVoltageFreq>maxVoltage){
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/SpectrumAnalysis/KTAggregatedChannelOptimizer.cc]: Fixed in commit 77127bb

File: Source/SpectrumAnalysis/KTAggregatedChannelOptimizer.hh [hunk 15]
```diff
@@ -0,0 +1,72 @@
+/*
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/SpectrumAnalysis/KTAggregatedChannelOptimizer.hh]: Fixed in commit 77127bb

File: Source/SpectrumAnalysis/KTChannelAggregator.cc [hunk 16]
```diff
@@ -0,0 +1,143 @@
+/*
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/SpectrumAnalysis/KTChannelAggregator.cc]: Fixed in commit 77127bb

File: Source/SpectrumAnalysis/KTChannelAggregator.cc [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+    // Assume a square grid. i.e, number of points in X= no of points in Y
+    KTAggregatedFrequencySpectrumDataFFTW& newAggFreqData=fftwData.Of< KTAggregatedFrequencySpectrumDataFFTW >().SetNComponents(fNGrid*fNGrid);
+    
+    // Setting up the active radius of the KTAggregatedFrequencySpectrumDataFFTW object to maintain consistency
+    // This doesn't need to be done if there is a way to provide config values to data objects
+    newAggFreqData.SetActiveRadius(fActiveRadius);
+    
+    int nTotalGridPoints=0;
+    // Loop over the grid points and fill the values
+    for (int iGridX=0; iGridX<fNGrid; iGridX++) {
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/SpectrumAnalysis/KTChannelAggregator.cc]: Fixed in commit 77127bb

File: Source/SpectrumAnalysis/KTChannelAggregator.hh [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+     Get radius of the active area.
+     */
+    virtual double GetActiveRadius() const;
+    
+    bool GetIsGridDefined() const;
+  private:
+    
+    // in meters, should not be hard-coded
+    // PTS: Has to come as an input from config file ?
+    double fActiveRadius = 0.0516;
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/SpectrumAnalysis/KTChannelAggregator.hh]: Fixed in commit 77127bb

File: Source/SpectrumAnalysis/KTChannelAggregator.hh (hunk 1 above)
Comment (psurukuc) [reply Source/SpectrumAnalysis/KTChannelAggregator.hh]: Not sure I understand the statement. Do you mean for me to change the argument in setter function for fIsGridDefined ? 
In commit 77127bb I used MEMBERVARIABLE() for fIsGridDefined. So, that might fix the issue ?

File: Source/SpectrumAnalysis/KTChannelAggregator.hh [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+     @class KTChannelAggregator
+     @author P. T. Surukuchi
+     
+     @brief Multiple channel summation for Phase-III and IV
+     
+     @details
+     More details to come.
+     
+     Configuration name: "channel-aggregator"
+
```
Comment (nsoblath) [inline Source/SpectrumAnalysis/KTChannelAggregator.hh]: Missing the documentation of the configuration values

File: Source/SpectrumAnalysis/KTChannelAggregator.cc [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+namespace Katydid
+{
+    KTLOGGER(agglog, "KTChannelAggregator");
+
+    // Register the processor
+    KT_REGISTER_PROCESSOR(KTChannelAggregator, "channel-aggregator");
+
+    KTChannelAggregator::KTChannelAggregator(const std::string& name) :
+            KTProcessor(name),
+            fPhaseChFrequencySumSlot("fft", this, &KTChannelAggregator::SumChannelVoltageWithPhase, &fSummedFrequencyData),
```
Comment (nsoblath) [inline Source/SpectrumAnalysis/KTChannelAggregator.cc]: The order of the initialization list needs to match the order in which member variables are defined.  Otherwise the compiler complains.

File: Source/SpectrumAnalysis/KTChannelAggregator.hh [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+        MEMBERVARIABLE(double, Wavelength);
+
+            //For exception handling to make sure the grid is defined before the spectra are assigned.
+        MEMBERVARIABLE(bool, IsGridDefined);
+
+        private:
+            bool SumChannelVoltageWithPhase(KTFrequencySpectrumDataFFTW&);
+
+            /// Returns the phase shift based on a given point, angle of the channel and the wavelength
+            double GetPhaseShift(double, double, double, double) const;
```
Comment (nsoblath) [inline Source/SpectrumAnalysis/KTChannelAggregator.hh]: For this and all functions with parameters, please give the parameter variable names in the header, so that a user can understand the meanings of the parameters just from looking at function definitions.

File: Source/IO/ROOTSpectrogramWriter/KTROOTSpectrogramWriter.hh [hunk 22]
```diff
@@ -72,14 +71,19 @@ namespace Katydid
             template< typename XDataType >
             void AddPowerSpectralDensityDataCoreHelper(Nymph::KTDataPtr data, DataTypeBundle& dataBundle);
 
+            template< typename XDataType >
+            void AddAggregatePowerSpectrumDataCoreHelper(Nymph::KTDataPtr data, DataTypeBundle& dataBundle);
```
Comment (nsoblath) [inline Source/IO/ROOTSpectrogramWriter/KTROOTSpectrogramWriter.hh]: Why did you need to add new helper functions?  These appear to be the same as the non-aggregated ones.

File: Source/SpectrumAnalysis/KTChannelAggregator.hh (hunk 21 above)
Comment (psurukuc) [reply Source/SpectrumAnalysis/KTChannelAggregator.hh]: Fixed in commit 8fea59e0bfb1ea3ac6f519d00ff17095c12cdddb

File: Source/Data/SpectrumAnalysis/KTChannelAggregatedData.hh [hunk 23]
```diff
@@ -0,0 +1,217 @@
...
+namespace Katydid
+{
+    class KTAggregatedDataCore
+    {
+        public:
+            KTAggregatedDataCore();
+
+            virtual ~KTAggregatedDataCore();
+
+            virtual void SetNGridPoints(unsigned);
```
[the remark this answers is not in the document]
Comment (psurukuc) [reply Source/Data/SpectrumAnalysis/KTChannelAggregatedData.hh]: Fixed in commit 4c724b3980ea5e75cd2aae460f3f70d5f1a43e97

File: Source/SpectrumAnalysis/KTChannelAggregator.hh (hunk 19 above)
Comment (psurukuc) [reply Source/SpectrumAnalysis/KTChannelAggregator.hh]: Fixed in commit ae569a85bc54b10279dfe0c08acefb98f8c59bd2

File: Source/IO/ROOTSpectrogramWriter/KTROOTSpectrogramWriter.hh (hunk 22 above)
Comment (psurukuc) [reply Source/IO/ROOTSpectrogramWriter/KTROOTSpectrogramWriter.hh]: This was a repetition in code that was caused because of a conflict with the develop branch. I have removed the repeated code. Fixed in commit ae569a85bc54b10279dfe0c08acefb98f8c59bd2.

File: Source/SpectrumAnalysis/KTChannelAggregator.cc (hunk 20 above)
Comment (psurukuc) [reply Source/SpectrumAnalysis/KTChannelAggregator.cc]: Fixed in 8fea59e0bfb1ea3ac6f519d00ff17095c12cdddb.
